modules/katello_activation_key.py

```diff
@@ -256,39 +157,54 @@ def main():
     if has_import_error:
         module.fail_json(msg=import_error_msg)
 
-    server_url = module.params['server_url']
-    username = module.params['username']
-    password = module.params['password']
-    verify_ssl = module.params['verify_ssl']
-    name = module.params['name']
-    organization = module.params['organization']
-    lifecycle_environment = module.params['lifecycle_environment']
-    content_view = module.params['content_view']
-    subscriptions = module.params['subscriptions']
-    auto_attach = module.params['auto_attach']
-    new_name = module.params['new_name']
-    state = module.params['state']
+    entity_dict = dict(
+        [(k, v) for (k, v) in module.params.items() if v is not None])
 
-    server = ServerConfig(
-        url=server_url,
-        auth=(username, password),
-        verify=verify_ssl
-    )
-    ng = NailGun(server, entities, module)
+    server_url = entity_dict.pop('server_url')
+    username = entity_dict.pop('username')
+    password = entity_dict.pop('password')
+    verify_ssl = entity_dict.pop('verify_ssl')
+    state = entity_dict.pop('state')
 
-    # Lets make an connection to the server with username and password
     try:
-        org = entities.Organization(server)
-        org.search()
+        create_server(server_url, (username, password), verify_ssl)
     except Exception as e:
         module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
 
+    ping_server(module)
+
+    entity_dict['organization'] = find_organization(module, name=entity_dict['organization'])
+    if entity_dict['lifecycle_environment']:
+        entity_dict['lifecycle_environment'] = find_lifecycle_environment(module, entity_dict['lifecycle_environment'], entity_dict['organization'])
+
+    if entity_dict['content_view']:
+        entity_dict['content_view'] = find_content_view(module, entity_dict['content_view'], entity_dict['organization'])
+
+    activation_key_dict = sanitize_entity_dict(entity_dict, name_map)
+    activation_key_entity = find_activation_key(module, name=entity_dict['name'], organization=entity_dict['organization'],
+                                                failsafe=True)
+
     try:
-        if state == 'copied':
-            changed = ng.activation_key_copy(name, organization, new_name)
-        else:
-            changed = ng.activation_key(name, organization, lifecycle_environment=lifecycle_environment, content_view=content_view,
-                                        subscriptions=subscriptions, auto_attach=auto_attach)
+        changed, activation_key_entity = naildown_entity(ActivationKey, activation_key_dict, activation_key_entity, state, module)
+
+        if state != 'copied':
```

Comment:
Otherwise it would require us to pass all subs to the `copied` invocation, making the copy less useful.

Reply:
Sounds reasonable, but i would not understand from reading the source alone. Can you please elaborate on this in a comment.

Reply:
Done, also changed to `state == present`, as otherwise it'd try to update subs of `state == absent` keys ;)